lib/rspec/support/method_signature_verifier.rb

```diff
@@ -154,19 +154,57 @@ def classify_parameters
       INFINITY = 1 / 0.0
     end
 
-    # Some versions of JRuby have a nasty bug we have to work around :(.
-    # https://github.com/jruby/jruby/issues/2816
+    # JRuby has only partial support for UnboundMethod#parameters, so we fall back on using #arity
+    # https://github.com/jruby/jruby/issues/2816 and https://github.com/jruby/jruby/issues/2817
     if RSpec::Support::Ruby.jruby? &&
        RubyFeatures.optional_and_splat_args_supported? &&
-       Class.new { attr_writer :foo }.instance_method(:foo=).parameters == []
+       Java::JavaLang::String.instance_method(:char_at).parameters == []
+      class MethodSignature < remove_const(:MethodSignature)
+      private
+
+        def classify_parameters
+          super
+          if (arity = @method.arity) != 0 && @method.parameters.empty?
+            if arity < 0
+              @min_non_kw_args = ~arity
+              @max_non_kw_args = INFINITY
+            else
+              @min_non_kw_args = arity
+              @max_non_kw_args = arity
+            end
+          end
+        end
+      end
+    end
 
+    if RSpec::Support::Ruby.jruby? &&
```

Comment:
Instead of starting a 2nd conditional that starts with `if RSpec::Support::Ruby.jruby?`, would you mind doing one big `if RSpec::Support::Ruby.jruby?` conditional (with 2 nested conditionals as needed).

Reply:
Actually, is it possible to combine the two Jruby cases into one class, with some extra conditional checked when `classify_parameters` is called? It's confusing (and will be harder to maintain) to have two different alternate JRuby implementations.

Reply:
The motivation for having them separate is that the latter one cannot use `#parameters`, but it could obviously be implemented by checking support for `#parameters` at runtime, if you prefer such a solution.